In [ ]:
import pandas as pd
school_datapath = "Resources/schools_complete.csv"
student_datapath= "Resources/students_complete.csv"

school_data = pd.read_csv(school_datapath)
student_data = pd.read_csv(student_datapath)

complete_school_data = pd.merge(student_data,school_data, how="left", on = ["school_name"])

In [2]:
#District Summary

#Setting up dictionary to make into a dataframe
series_names = {}
#Setting up dataframe with students who passed both math and reading
passingboth_df = complete_school_data.loc[(complete_school_data["math_score"] >= 70) & (complete_school_data["reading_score"] >=  70), :]

    
#Total Number of Schools calculation
series_names["Total Schools"] = len(complete_school_data["school_name"].unique())
#Total Number of Students calculation
series_names["Total Students"] = "{:,.0f}".format(complete_school_data["Student ID"].count())
#Total Budget calculation w/ 2 decimal formatting
series_names["Total Budget"] = "${:,.2f}".format(school_data["budget"].sum())
#Average Math Score calculation w/ percent formatting
series_names["Average Math Score"] = "{:,.2f}%".format(complete_school_data["math_score"].mean())
#Average Reading Score calculation w/ percent formatting
series_names["Average Reading Score"] = "{:,.2f}%".format(complete_school_data["reading_score"].mean())
#Percentage of students passing math
passing_math = (complete_school_data["math_score"] >= 70).value_counts()
series_names["% Passing Math"] = "{:,.2f}%".format(passing_math[True] * 100/ (passing_math[True] + passing_math[False]))
#Percentage of students passing reading
passing_reading = (complete_school_data["reading_score"] >= 70).value_counts()
series_names["% Passing Reading"] = "{:,.2f}%".format(passing_reading[True] * 100/ (passing_reading[True] + passing_reading[False]))
#Overall % of students passing
series_names["% Overall Passing"] = "{:,.2f}%".format(len(passingboth_df)*100/((passing_math[True] + passing_math[False])))
#Dictionary to dataframe conversion
series_names_df = pd.DataFrame(series_names, index = [0])
series_names_df.head()









Total Schools Total Students    Total Budget Average Math Score  \
0             15         39,170  $24,649,428.00             78.99%   

  Average Reading Score % Passing Math % Passing Reading % Overall Passing  
0                81.88%         74.98%            85.81%            65.17%

In [38]:
#School District Summary
# Formatting for later  ["${:,.2f}".format(number) for number in df(replace with dataframe series name)]
school_names = complete_school_data["school_name"].unique()
school_math_averages = {}
school_reading_averages = {}
school_passing_math = {}
school_passing_reading = {}
school_passing_overall = {}
school_summary_df = pd.DataFrame(index = school_names)
school_summary_df["School Type"] = school_data["type"].tolist()
school_summary_df["Total Students"] = school_data["size"].tolist()
school_summary_df["Total Budget"] = ["${:,.2f}".format(number) for number in school_data["budget"].tolist()]


school_summary_df["Per Student Budget"] = (school_data["budget"] / school_data["size"]).tolist()

#Itearting through the series and testing a conditional to return score if school names match; stored in a dictionary
for name in school_names:
   school_math_averages[name] = (student_data.loc[student_data["school_name"] == name, "math_score"]).mean()

for name in school_names:
   school_reading_averages[name] = (student_data.loc[student_data["school_name"] == name, "reading_score"].mean())

school_summary_df["Average Math Score"] = school_math_averages.values()
school_summary_df["Average Reading Score"] = school_reading_averages.values()

#Iterating over a dataframe to test for passing math score; store results in dictionary for later
for name in school_names:
   school = student_data[["math_score","reading_score"]][student_data["school_name"] == name]
   school_passing = (school["math_score"] >= 70).value_counts()
   school_passing_math[name] = school_passing[True] * 100 / len(school)

#Iterating over a dataframe to test for passing reading score; store results in dictionary for later
for name in school_names:
   school = student_data[["math_score","reading_score"]][student_data["school_name"] == name]
   school_passing = (school["reading_score"] >= 70).value_counts()
   school_passing_reading[name] = school_passing[True] * 100 / len(school)

#Iterating over a dataframe to test for passing scores in both math/reading. Apparently there was an easier way to do this using groupby()...   
for name in school_names:
   school = student_data[["math_score","reading_score"]][student_data["school_name"] == name]
   school_passing = ((school["reading_score"] >= 70) & (school["math_score"] >= 70)).value_counts()
   school_passing_overall[name] = school_passing[True] * 100 / len(school)

#Formatting the series in the data frames.
school_summary_df["% Passing Math"] = school_passing_math.values()
school_summary_df["% Passing Reading"] = school_passing_reading.values()
school_summary_df["% Overall Passing"] = school_passing_overall.values()

#Sorted Dataframe based on sized in ascending order
school_summary_df.sort_values("Total Students")

School Type  Total Students   Total Budget  \
Holden High School        Charter             427    $248,087.00   
Pena High School          Charter             962    $585,858.00   
Griffin High School       Charter            1468    $917,500.00   
Thomas High School        Charter            1635  $1,043,130.00   
Shelton High School       Charter            1761  $1,056,600.00   
Wright High School        Charter            1800  $1,049,400.00   
Cabrera High School       Charter            1858  $1,081,356.00   
Wilson High School        Charter            2283  $1,319,574.00   
Ford High School         District            2739  $1,763,916.00   
Huang High School        District            2917  $1,910,635.00   
Figueroa High School     District            2949  $1,884,411.00   
Rodriguez High School    District            3999  $2,547,363.00   
Hernandez High School    District            4635  $3,022,020.00   
Johnson High School      District            4761  $3,094,650.00   
Bailey High School       District            4976  $3,124,928.00   

                       Per Student Budget  Average Math Score  \
Holden High School                  581.0           83.803279   
Pena High School                    609.0           83.839917   
Griffin High School                 625.0           83.351499   
Thomas High School                  638.0           83.418349   
Shelton High School                 600.0           83.359455   
Wright High School                  583.0           83.682222   
Cabrera High School                 582.0           83.061895   
Wilson High School                  578.0           83.274201   
Ford High School                    644.0           77.102592   
Huang High School                   655.0           76.629414   
Figueroa High School                639.0           76.711767   
Rodriguez High School               637.0           76.842711   
Hernandez High School               652.0           77.289752   
Johnson High School                 650.0           77.072464   
Bailey High School                  628.0           77.048432   

                       Average Reading Score  % Passing Math  \
Holden High School                 83.814988       92.505855   
Pena High School                   84.044699       94.594595   
Griffin High School                83.816757       93.392371   
Thomas High School                 83.848930       93.272171   
Shelton High School                83.725724       93.867121   
Wright High School                 83.955000       93.333333   
Cabrera High School                83.975780       94.133477   
Wilson High School                 83.989488       93.867718   
Ford High School                   80.746258       68.309602   
Huang High School                  81.182722       65.683922   
Figueroa High School               81.158020       65.988471   
Rodriguez High School              80.744686       66.366592   
Hernandez High School              80.934412       66.752967   
Johnson High School                80.966394       66.057551   
Bailey High School                 81.033963       66.680064   

                       % Passing Reading  % Overall Passing  
Holden High School             96.252927          89.227166  
Pena High School               95.945946          90.540541  
Griffin High School            97.138965          90.599455  
Thomas High School             97.308869          90.948012  
Shelton High School            95.854628          89.892107  
Wright High School             96.611111          90.333333  
Cabrera High School            97.039828          91.334769  
Wilson High School             96.539641          90.582567  
Ford High School               79.299014          54.289887  
Huang High School              81.316421          53.513884  
Figueroa High School           80.739234          53.204476  
Rodriguez High School          80.220055          52.988247  
Hernandez High School          80.862999          53.527508  

In [14]:
#Top 5 schools

sorted_schools_df = school_summary_df.sort_values(by="% Overall Passing", ascending=False)
top5_df = sorted_schools_df.iloc[:5]
top5_df

School Type  Total Students   Total Budget  \
Cabrera High School     Charter            1858  $1,081,356.00   
Thomas High School      Charter            1635  $1,043,130.00   
Griffin High School     Charter            1468    $917,500.00   
Wilson High School      Charter            2283  $1,319,574.00   
Pena High School        Charter             962    $585,858.00   

                    Per Student Budget Average Math Score  \
Cabrera High School            $582.00             83.06%   
Thomas High School             $638.00             83.42%   
Griffin High School            $625.00             83.35%   
Wilson High School             $578.00             83.27%   
Pena High School               $609.00             83.84%   

                    Average Reading Score % Passing Math % Passing Reading  \
Cabrera High School                83.98%         94.13%            97.04%   
Thomas High School                 83.85%         93.27%            97.31%   
Griffin High School                83.82%         93.39%            97.14%   
Wilson High School                 83.99%         93.87%            96.54%   
Pena High School                   84.04%         94.59%            95.95%   

                    % Overall Passing  
Cabrera High School            91.33%  
Thomas High School             90.95%  
Griffin High School            90.60%  
Wilson High School             90.58%  
Pena High School               90.54%

In [15]:
#Lowest Performing Schools
last5_df = sorted_schools_df[-5:]
last5_df

School Type  Total Students   Total Budget  \
Johnson High School      District            4761  $3,094,650.00   
Hernandez High School    District            4635  $3,022,020.00   
Huang High School        District            2917  $1,910,635.00   
Figueroa High School     District            2949  $1,884,411.00   
Rodriguez High School    District            3999  $2,547,363.00   

                      Per Student Budget Average Math Score  \
Johnson High School              $650.00             77.07%   
Hernandez High School            $652.00             77.29%   
Huang High School                $655.00             76.63%   
Figueroa High School             $639.00             76.71%   
Rodriguez High School            $637.00             76.84%   

                      Average Reading Score % Passing Math % Passing Reading  \
Johnson High School                  80.97%         66.06%            81.22%   
Hernandez High School                80.93%         66.75%            80.86%   
Huang High School                    81.18%         65.68%            81.32%   
Figueroa High School                 81.16%         65.99%            80.74%   
Rodriguez High School                80.74%         66.37%            80.22%   

                      % Overall Passing  
Johnson High School              53.54%  
Hernandez High School            53.53%  
Huang High School                53.51%  
Figueroa High School             53.20%  
Rodriguez High School            52.99%

In [20]:
#Math Scores by Grade

school_grades = ["9th","10th","11th","12th"]

math_by_grade_df = pd.DataFrame(index = school_names)

for grade in school_grades:
    math_by_grade_df[grade] = (student_data[student_data["grade"] == grade]).groupby("school_name")["math_score"].mean()
    math_by_grade_df[grade] = ["{:,.2f}%".format(number) for number in math_by_grade_df[grade]]
math_by_grade_df


9th    10th    11th    12th
Huang High School      77.03%  75.91%  76.45%  77.23%
Figueroa High School   76.40%  76.54%  76.88%  77.15%
Shelton High School    83.42%  82.92%  83.38%  83.78%
Hernandez High School  77.44%  77.34%  77.14%  77.19%
Griffin High School    82.04%  84.23%  83.84%  83.36%
Wilson High School     83.09%  83.72%  83.20%  83.04%
Cabrera High School    83.09%  83.15%  82.77%  83.28%
Bailey High School     77.08%  77.00%  77.52%  76.49%
Holden High School     83.79%  83.43%  85.00%  82.86%
Pena High School       83.63%  83.37%  84.33%  84.12%
Wright High School     83.26%  84.01%  83.84%  83.64%
Rodriguez High School  76.86%  76.61%  76.40%  77.69%
Johnson High School    77.19%  76.69%  77.49%  76.86%
Ford High School       77.36%  77.67%  76.92%  76.18%
Thomas High School     83.59%  83.09%  83.50%  83.50%

In [21]:
#Reading Scores by Grade
reading_by_grade_df = pd.DataFrame(index = school_names)

for grade in school_grades:
    reading_by_grade_df[grade] = (student_data[student_data["grade"] == grade]).groupby("school_name")["reading_score"].mean()
    reading_by_grade_df[grade] = ["{:,.2f}%".format(number) for number in reading_by_grade_df[grade]]
reading_by_grade_df


9th    10th    11th    12th
Huang High School      81.29%  81.51%  81.42%  80.31%
Figueroa High School   81.20%  81.41%  80.64%  81.38%
Shelton High School    84.12%  83.44%  84.37%  82.78%
Hernandez High School  80.87%  80.66%  81.40%  80.86%
Griffin High School    83.37%  83.71%  84.29%  84.01%
Wilson High School     83.94%  84.02%  83.76%  84.32%
Cabrera High School    83.68%  84.25%  83.79%  84.29%
Bailey High School     81.30%  80.91%  80.95%  80.91%
Holden High School     83.68%  83.32%  83.82%  84.70%
Pena High School       83.81%  83.61%  84.34%  84.59%
Wright High School     83.83%  83.81%  84.16%  84.07%
Rodriguez High School  80.99%  80.63%  80.86%  80.38%
Johnson High School    81.26%  80.77%  80.62%  81.23%
Ford High School       80.63%  81.26%  80.40%  80.66%
Thomas High School     83.73%  84.25%  83.59%  83.83%

In [39]:
#Scores by School Spending
school_spending_df = school_summary_df[["Average Math Score", "Average Reading Score", "% Passing Math","% Passing Reading", "% Overall Passing"]]
sorted_spending_df = school_spending_df.sort_values("% Overall Passing", ascending = False)
sorted_spending_df["$ Spent Per Student"] = pd.cut(school_summary_df["Per Student Budget"], [578,585,630,645,655], include_lowest=True, labels = ["<$585","$585-$630","$630-$645","$655-$680"])
final_sorted_spending_df = sorted_spending_df.groupby("$ Spent Per Student").mean()
final_sorted_spending_df




Average Math Score  Average Reading Score  \
$ Spent Per Student                                              
<$585                         83.455399              83.933814   
$585-$630                     81.899826              83.155286   
$630-$645                     78.518855              81.624473   
$655-$680                     76.997210              81.027843   

                     % Passing Math  % Passing Reading  % Overall Passing  
$ Spent Per Student                                                        
<$585                     93.460096          96.610877          90.369459  
$585-$630                 87.133538          92.718205          81.418596  
$630-$645                 73.484209          84.391793          62.857656  
$655-$680                 66.164813          81.133951          53.526855

In [40]:
#School Performance by size
school_size_df = school_summary_df[["Average Math Score", "Average Reading Score","% Passing Math","% Passing Reading", "% Overall Passing"]]
sorted_size_df = school_size_df.sort_values("% Overall Passing", ascending = False)
sorted_size_df["School Size"] = pd.cut(school_summary_df["Total Students"], [427,1000,2000,5000], include_lowest = True, labels = ["Small <1000", "Medium 1000-2000", "Large 2000-5000"])
final_sorted_size_df = sorted_size_df.groupby("School Size").mean()
final_sorted_size_df

Average Math Score  Average Reading Score  % Passing Math  \
School Size                                                                   
Small <1000                83.821598              83.929843       93.550225   
Medium 1000-2000           83.374684              83.864438       93.599695   
Large 2000-5000            77.746417              81.344493       69.963361   

                  % Passing Reading  % Overall Passing  
School Size                                             
Small <1000               96.099437          89.883853  
Medium 1000-2000          96.790680          90.621535  
Large 2000-5000           82.766634          58.286003

In [41]:
#District Type
school_type_df = school_summary_df[["Average Math Score", "Average Reading Score","% Passing Math","% Passing Reading", "% Overall Passing"]]
sorted_type_df = school_type_df.sort_values("% Overall Passing", ascending = False)
school_type_map = {"District":0, "Charter":1}
school_type_temp = (school_summary_df["School Type"]).map(school_type_map)

sorted_type_df["School Type"] = pd.cut(school_type_temp, 2, labels = ["District","Charter"])
final_sorted_type_df = sorted_type_df.groupby("School Type").mean()
final_sorted_type_df

Average Math Score  Average Reading Score  % Passing Math  \
School Type                                                              
District              76.956733              80.966636       66.548453   
Charter               83.473852              83.896421       93.620830   

             % Passing Reading  % Overall Passing  
School Type                                        
District             80.799062          53.672208  
Charter              96.586489          90.432244

<strong> School District Report </strong>

After consolidation of the school data set and student data set, there are a few observations to be made. The following observations can be made; school type and school size are two key factors in the overall passing percentage of students. School type refers to whether or not a school is considered a Charter or a District school. According to the data, **charter schools are more likely to have a higher overall percentage of students that are passing both math and reading**. Data shows that charter schools have 90.43% of their student population passing both math and reading whereas district schools only have 53.67% of their student population passing both subjects. 

This leads into the next observation, where by looking at whether or not a school is considered a District or Charter school one can also reliably guess at the size of the school. District schools tend to be bigger than charter schools, in fact comparing the smallest district schools vs smallest charter schools one can see that the smallest district school only consists of 427 students whereas the smallest district school consists of over 2000 students. **These sizes directly correlate with overall passing rate as there is an immediate decline from 90.58% of the highest populated charter (2283) to the lowest populated district school (2739) with a overall passing rate of 54.29%**. 

Moving forward, the school district may want to incentivize the distribution of school populations. Of course, there will be logistic difficulties regarding distance to schools and the school facility size itself among other things. However, we as a district should find a way to rebalance how many students are allocated to each school as a way to bolster passing rates and provide the best education to our students and their families. 

-	Davit Ksor 09/2022
